spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,147 @@
+import org.springframework.integration.endpoint.AbstractFetchLimitingMessageSource;
+import org.springframework.integration.endpoint.AbstractMessageSource;
+import org.springframework.integration.mapping.HeaderMapper;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+/**
+ * A <a href="https://cloud.google.com/pubsub/docs/pull#pubsub-pull-messages-sync-java">PubSub Synchronous pull</a>
+ * implementation of {@link AbstractMessageSource}.
+ *
+ * <p>Enables more even load balancing with polling adapters, at the cost of efficiency gained from streaming pull.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.2
+ */
+public class PubSubMessageSource extends AbstractFetchLimitingMessageSource<Object> {
+
+	private final String subscriptionName;
+
+	private final PubSubSubscriberOperations pubSubSubscriberOperations;
+
+	private AckMode ackMode = AckMode.MANUAL;
+
+	private HeaderMapper<Map<String, String>> headerMapper = new PubSubHeaderMapper();
+
+	private Class payloadType = byte[].class;
+
+	private boolean blockOnPull;
+
+	private ArrayDeque<ConvertedAcknowledgeablePubsubMessage> cachedMessages = new ArrayDeque<>();
+
+	public PubSubMessageSource(PubSubSubscriberOperations pubSubSubscriberOperations,
+			String subscriptionName) {
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		Assert.notNull(ackMode, "The acknowledgement mode can't be null.");
+		this.ackMode = ackMode;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		Assert.notNull(payloadType, "The payload type cannot be null.");
+		this.payloadType = payloadType;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		Assert.notNull(headerMapper, "The header mapper can't be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	/**
+	 * Instructs synchronous pull to wait until at least one message is available.
+	 */
+	public void blockOnPull() {
+		this.blockOnPull = true;
+	}
+
+	/**
+	 * Provides a single polled message.
+	 * <p>Messages are received from Pub/Sub by synchronous pull, in batches determined
+	 * by {@code fetchSize}.
+	 * @param fetchSize number of messages to fetch from Pub/Sub.
+	 * @return {@link Message} wrapper containing the original message.
+	 */
+	@Override
+	protected Object doReceive(int fetchSize) {
+		if (this.cachedMessages.isEmpty()) {
+			Integer maxMessages = (fetchSize > 0) ? fetchSize : 1;
+
+			List<ConvertedAcknowledgeablePubsubMessage> messages
+					= this.pubSubSubscriberOperations.pullAndConvert(this.subscriptionName, maxMessages, !this.blockOnPull, this.payloadType);
+			if (messages.isEmpty()) {
+				return null;
+			}
+			else if (messages.size() == 1) {
```

Comment:
I would just skip this branch because the `else` adequately handles this case.

Reply:
I don't know; putting a message into a data structure (no matter how efficient) only to immediately take it out does not sit well with me in something that is as high processing volume as pubsub.

Reply:
Probably negligible performance impact considering everything else that needs to happen up to this point, but you have a point.